Src/FluentAssertions/Formatting/AttributeBasedFormatter.cs

```diff
@@ -46,11 +47,45 @@ public string Format(object value, FormattingContext context, FormatChild format
         private MethodInfo GetFormatter(object value)
         {
             Type valueType = value.GetType();
-            MethodInfo formatter = Formatters.FirstOrDefault(m => m.GetParameters().Single().ParameterType == valueType);
+            if (!determinedFormatters.TryGetValue(valueType, out var formatter))
+            {
+                var possibleFormatters = Formatters
+                    .Select(m => new { Type = m.GetParameters().Single().ParameterType, Method = m })
+                    .GroupBy(f => f.Type)
+                    .Select(g => g.First())
+                    .Where(f => valueType.IsSameOrInherits(f.Type))
+                    .ToDictionary(f => f.Type, f => f.Method);
+
+                formatter = GetFormatter(possibleFormatters, valueType);
+                determinedFormatters.Add(valueType, formatter);
```

Comment:
Sorry for being too vague on my thoughts about caching.

As `AttributeBasedFormatter` is instantiated a single time in the static `Formatter.DefaultFormatters` we cannot use `Dictionary<TKey, TValue>` as a cache as `Add` is not thread-safe.

My idea is to change 
```cs
private MethodInfo[] formatters;
``` 
into
```cs
private Dictionary<Type, MethodInfo> formatters;
```

```cs
private Dictionary<Type, MethodInfo> Formatters
{
    get
    {
        HandleValueFormatterDetectionModeChanges();

        return formatters ??= GetDictionary(FindCustomFormatters());
    }
}

// with a better name or merge into FindCustomFormatters
private static Dictionary<Type, MethodInfo> GetDictionary(MethodInfo[] formatters)
{
    return formatters
        .Select(m => new { Type = m.GetParameters().Single().ParameterType, Method = m })
        .GroupBy(f => f.Type)
        .Select(g => g.First())
        .ToDictionary(f => f.Type, f => f.Method);
}
```

```cs
private MethodInfo GetFormatter(object value)
{
    Type valueType = value.GetType();
    return GetFormatter(valueType);
}

private MethodInfo GetFormatter(Type valueType)
{
    do
    {
        if (Formatters.TryGetValue(valueType, out var formatter))
        {
            return formatter;
        }

        valueType = valueType.BaseType;
    }
    while (valueType != null);

    return null;
}
```

Reply:
Thanks @jnyrup, got it.
You miss `IsSameOrInherits` check in your example. How do you suggest including it?

Reply:
I deliberately ignored `IsSameOrInherits` 😀 
I might be missing something, but I couldn't see it was needed?

Reply:
There was a glitch in my head, I was confused by `possibleFormatters.Count == 1` when we didn't need to analyze type. But I don't think this is really needed. Finally pushed your example.
Just wondering why AttributeBasedFormatter is public.